refactor(api,robot-server): Expose protocol metadata to HTTP clients

SyntaxColoring opened the pull request:
# Overview

This PR exposes protocol metadata (like author name and protocol description) through the HTTP API. Closes #8003.

# Code review requests

* Anyone have a better term for this than "pre-analysis"?
* Check out the `protocol_models.Metadata` docstring. Do we agree on this rough definition of metadata, from the perspective of integrators and protocol authors? Further specification is ticketed as #8334.
* Are we still sure about committing to metadata always being available immediately at upload time? Alternatively, the API for reading metadata could work the same way as our API for getting required equipment: a background task could start in the background when you upload the protocol, and you could fetch its results when it finishes, later. (Which might be in 0.1 seconds, or it might be in 30 seconds.)

# Manual test plan

I've done this:

1. Enable the **Enable experimental Protocol Engine** feature flag.
2. `POST /protocols` with a JSON or Python protocol. @mcous's [Postman HTTP collection](https://gist.github.com/mcous/7dde93b6ea83753b86b0d84b40ba07f6) can help.
3. Confirm that the protocol's metadata appears in the response.
    * All of the protocol's metadata fields should be passed straight through. Their types should be retained.
    * There should be no fields in the response that the protocol did *not* explicitly set.
4. Try some error cases:
    * Multi-file protocols should currently raise a `NotImplementedError`
    * Other problems with parsing or analysis should respond with an HTTP 400

This should work with `make dev` as long as you have the right feature flags set; you don't need a real robot.

# Risk assessment

Low. This only touches stuff in Protocol Engine and the new HTTP API.

SyntaxColoring committed:
Flesh out `Metadata` docstring.

SyntaxColoring committed:
Remove `pythonApiLevel` from response model.

SyntaxColoring committed:
WIP: Return a protocol's metadata on upload.

SyntaxColoring committed:
Lint.

SyntaxColoring committed:
Make it mandatory to explicitly specify `Protocol.protocolMetadata`.

SyntaxColoring committed:
Fix docstring: `metadata` -> `protocolMetadata`

SyntaxColoring committed:
WIP: Update tests

SyntaxColoring committed:
Fix `ProtocolStore` test passing an already-iterated file.

SyntaxColoring committed:
Revert most test changes.

SyntaxColoring committed:
Revert `ProtocolStore` changes.

SyntaxColoring committed:
Merge branch 'edge' into pe_respond_with_metadata

SyntaxColoring committed:
Update `opentrons.protocol_runner.__init__.py`.

SyntaxColoring committed:
Update `ResponseBuilder` test.

SyntaxColoring committed:
WIP: Work on typechecking and linting.

SyntaxColoring committed:
Fix up import and use-before-declaration.

SyntaxColoring committed:
Dependency-inject a `PreAnalyzer` into `ProtocolStore`.

SyntaxColoring committed:
Appease the linter.

SyntaxColoring committed:
Add skeleton for `.analyze()`.

SyntaxColoring committed:
Move `_make_json_protocol()` test helper to `models.json_protocol`.

SyntaxColoring committed:
Make `metadata` customizable in `make_minimal()`.

SyntaxColoring committed:
Add some `PreAnalyzer` tests.

mcous commented on api/src/opentrons/protocol_runner/__init__.py:
```diff
@@ -5,11 +5,15 @@
 
 __all__ = [
     "ProtocolRunner",
     "ProtocolFile",
```
I think this PR is a good time to rename this to `ProtocolSource`

SyntaxColoring replied:
Will rename to `ProtocolRunSource` because `ProtocolSource` sounds to me like it could be the input to the pre-analyzer.

SyntaxColoring replied:
[Discussed in call] Leaving this alone for now until after refactor to put stuff on the disk and deal with `pathlib.Path` objects earlier. (See `todo` comment in code.)

mcous commented on robot-server/robot_server/protocols/protocol_models.py:
```diff
@@ -1,23 +1,71 @@
+
+    class Config:  # noqa: D106
+        # Tell Pydantic that objects of this model can have arbitrary fields.
+        extra = "allow"
```
Should we just use a `Dict[str, Any]` rather than having a `BaseModel` with only arbitrary fields?

SyntaxColoring replied:
Maybe, yeah. I'll investigate and see what produces a nicer OpenAPI spec.

mcous replied:
Let's use the enum provided by Pydantic instead

```py
from pydantic import BaseModel, Field, Extra
# ...
extra = Extra.allow

SyntaxColoring replied:
Changed to use enum.

Deferring the `Dict[str, Any]` thing.

mcous commented on robot-server/robot_server/protocols/dependencies.py:
```diff
@@ -26,7 +26,7 @@ def get_protocol_store(app_state: AppState = Depends(get_app_state)) -> Protocol
     if protocol_store is None:
         log.info(f"Storing protocols in {_PROTOCOL_STORE_DIRECTORY}")
-        protocol_store = ProtocolStore(directory=_PROTOCOL_STORE_DIRECTORY)
+        protocol_store = ProtocolStore(_PROTOCOL_STORE_DIRECTORY, PreAnalyzer())
```
You might have already gotten here yourself, but I would not inject `PreAnalyzer` into the `ProtocolStore`. The store is currently not a collaborator, and if you inject `PreAnalyzer`, you'll turn it into one. From there, you'll either 

1. Have redundant tests if you continue to opaquely test `ProtocolStore`
2. Need to make another "pure" unit to handle the storage mechanisms of `ProtocolStore`, testing the "new" collaborator `ProtocolStore` using mocks

Neither of those sound fun to me. Instead, looking at our current / potential units in play:

- `get_protocol` route handler - collaborator
- `ProtocolStore` - "pure" unit
    - Takes vanilla files
    - Copies files to disk
    - Creates protocol resources
    - Stores protocol resources
- `PreAnalyzer` - "pure" unit
    - Takes vanilla files
    - ???

Looking at that, I think one collaborator (the route handler) is still good, so I might end up shifting these responsibilities around. Hopefully that helps with naming, too:

- `PreAnalyzer`
    - Takes vanilla file-like's
        - [Starlette `UploadFiles` are conveniently file-like](https://www.starlette.io/requests/#request-files), and will have the file contents spooled up in memory
    - File reading stuff to parse metadata, determine protocol type, etc.
    - Returns a data structure that includes the file-like objects and parsed stuff
- `ProtocolStore`
    - Takes whatever the `PreAnalyzer` returns
    - Puts files on disk
    - Returns `ProtocolSource`
- Route handler
    - Passes files to `PreAnalyzer`
    - Passes `PreAnalyzer` return value to `ProtocolStore`
    - Passes `ProtocolStore` return value to response

mcous reviewed (commented):
I know this PR is still very in progress, but I have some thoughts about unit responsibilities that may help with the naming stuff request in the PR description, so here we go

SyntaxColoring committed:
Continue fleshing out tests.

* Implement tests for various errors.
* Add `JsonParseError` and `JsonSchemaValidationError` as distinct exception types.

SyntaxColoring committed:
Implement JSON pre-analysis and flesh out JSON tests.

SyntaxColoring commented on api/src/opentrons/protocol_runner/__init__.py:
```diff
@@ -5,11 +5,15 @@
     "ProtocolRunner",
     "ProtocolFile",
     "ProtocolFileType",
+    "PreAnalyzer",
+    "PythonPreAnalysis",
+    "JsonPreAnalysis",
```
If we're re-exporting the `pre_analysis` interface here, this should probably include error types, too.

If that becomes noisy and cumbersome, maybe don't reexport anything, and expect callers to import the `protocol_runner.pre_analysis` module to be imported as its own thing.

SyntaxColoring commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,85 @@
+from opentrons.protocols.models import JsonProtocol
+
+
+class PreAnalyzer:  # noqa: D101
+    @staticmethod
+    def analyze(  # noqa: D102
```
It's weird that Codecov is complaining about these lines not being covered by tests. Is something wrong with our PyTest coverage collection?

SyntaxColoring replied:
Local coverage report looks sensible. Theory: we're looking at alerts from the last CI job that passed. Tests are currently failing, so CI is failing, so these alerts haven't been updated.

SyntaxColoring commented on api/tests/opentrons/protocol_runner/test_pre_analysis.py:
```diff
@@ -0,0 +1,133 @@
+            tags=["tag1", "tag2"],
+            created=123.456,
+        )
+    ).dict(exclude_unset=True)
```
Investigate if we can set this in the config of the model

mcous replied:
As discussed, I don't think `exclude_unset` is worth the trouble here

SyntaxColoring replied:
> Investigate if we can set this in the config of the model

Turns out no, we can't. :(

> As discussed, I don't think `exclude_unset` is worth the trouble here

I'd like to keep it because:

* We just have to set this kwarg in a couple of places, so it's not out of our way.
* It matches the apparent intended behavior of our JSON protocol schema. (Acknowledging that we could just decree in #8334 that not-present is the same as `null. But I'd like to avoid making that decision right here and now.)
* It matches the intuitive meaning of "pass metadata straight-through without modification."
* It makes this test more readable and a bit less fragile. We don't have to fill `expected_metadata` with a bunch of `None`s, and add additional `None`s every time we add a new metadata field to the JSON protocol schema.

SyntaxColoring commented on api/tests/opentrons/protocol_runner/test_pre_analysis.py:
```diff
@@ -0,0 +1,133 @@
+            created=123.456,
+        )
+    ).dict(exclude_unset=True)
+    # exclude_unset=True prevents Pydantic from auto-filling metadata fields that the
+    # model has but that we omit here (like metadata.lastModified) with the value None.
+    #
+    # A caller could provide a protocol with fields omitted like this. We want to test
+    # that we preserve that omission, in case there's ever a meaningful difference
+    # between an omitted field and a null field.
+
+    # Give the input protocol a metadata field that our JSON protocol schema bindings
+    # don't know about. We want to test that we retain it, instead of silently
+    # discarding it. An old robot server, with old schema bindings, should be able to
+    # pass along metadata from new protocols even if it doesn't know about that
+    # metadata.
```
Maybe give this nuance its own test, separating it from the simple case.

SyntaxColoring replied:
Not bothering with this in this PR, unless someone wants me to.

SyntaxColoring committed:
Add even more exception types and cover them with tests.

SyntaxColoring committed:
Implement Python pre-analysis.

SyntaxColoring committed:
Remove reexports from __init__.py.

Keeping this list complete with all the exception types seemed untenable.
Better to just expect the pre_analysis module to be imported by name.

SyntaxColoring committed:
*Actually* remove reexports.

SyntaxColoring committed:
Add missing `pre_analysis` docstrings.

SyntaxColoring committed:
Leave error handling todo.

SyntaxColoring commented on api/tests/opentrons/protocol_runner/test_pre_analysis.py:
```diff
@@ -0,0 +1,228 @@
+    assert result == PythonPreAnalysis(expected_metadata, expected_api_level)
+
+
+# Apparently a bug in opentrons.protocols.parse.extract_metadata.
+# It errors when a field isn't a string, even though its return type suggests
+# it should allow ints.
```
Ticketing further investigation into

* what we currently actually allow
* what we should allow

as part of #8334.

SyntaxColoring committed:
Make `pre_analysis` take file-likes as input, not `pathlib.Path`s.

SyntaxColoring committed:
Add "raises" note to `ProtocolStore.create()` docstring.

SyntaxColoring committed:
Leave todo for closing `UploadFile`s.

SyntaxColoring committed:
Remove file quantity check from router.

This is covered by `PreAnalyzer` now.

SyntaxColoring committed:
Don't dependency-inject `PreAnalyzer` into `ProtocolStore`.

SyntaxColoring committed:
Update `protocol_store` tests.

SyntaxColoring committed:
Add missing `await` to `seek()` call.

SyntaxColoring committed:
Update `test_protocols_router.test_create_protocol()`.

SyntaxColoring committed:
Delete `test_protocols_router.test_create_multifile_protocol()`.

See 4e94e07c7f49acdd4c126e1279a34326ff9dd69e. This unit is not responsible for this anymore.

SyntaxColoring committed:
Format.

SyntaxColoring committed:
Test that router returns error if pre-analysis fails.

SyntaxColoring committed:
Update `test_protocols_router.test_create_protocol_invalid_file()`.

SyntaxColoring committed:
Lint.

SyntaxColoring commented on robot-server/tests/protocols/test_protocol_store.py:
```diff
@@ -147,43 +147,47 @@ async def test_get_missing_protocol_raises(
-            files=[matchers.Anything()],
-        ),
-    ]
+# async def test_get_all_protocols(
```
Reminder to self: uncomment/fix this.

SyntaxColoring committed:
Merge branch 'edge' into pe_respond_with_metadata

Conflicts:
* robot-server/robot_server/protocols/dependencies.py

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+
+def _analyze_json(main_file: InputFile) -> "JsonPreAnalysis":
+    try:
+        parsed_json = json_load(main_file.file_like)
```
Why not use `JsonProtocol.parse_file`? Also, shot in the dark, would the file seeking issues be resolved if we opened and closed the file-like?

```python
with open(main_file.file) as f:
    parsed_json = JsonProtocol.parse_file(f)
```

SyntaxColoring replied:
> Also, shot in the dark, would the file seeking issues be resolved if we opened and closed the file-like?

Good thought, but it turns out no, unfortunately.

A `SpooledTemporaryFile` can't be reopened after it's closed. This sort of makes sense: closing a temporary file is signaling that we no longer need it, so the library can then free the memory it was occupying or delete it from the filesystem. There might not be anything to reopen if we come back to it again.

mcous replied:
Cool cool cool

SyntaxColoring replied:
> Why not use `JsonProtocol.parse_file`?

A combination of a couple of reasons.

First, it turns out `.parse_file()` takes a *path,* not a file-like object. 🙃 

Second, even if it could take a file-like object, that would make it look something like this:

```python
    try:
        parsed_protocol = JsonProtocol.parse_file(main_file.file_like)
    except JSONDecodeError  as exception:
        raise JsonParseError() from exception
    except PydanticValidationError) as exception:
        # EDITED THIS CODE SAMPLE AFTER YOUR REPLY BELOW:
        # Separated this except block from the one above. Previous version of this code sample
        # raised JsonSchemaValidationError for both because I had a brain fart and forgot I defined
        # a separate JsonParseError().
        raise JsonSchemaValidationError() from exception
```

In other words we can't get rid of the `JSONDecodeError`. Pydantic doesn't appear to convert it to `PydanticValidationError`; it lets it propagate.

Catching it like this felt fragile to me because Pydantic's choice of default JSON deserializer is an undocumented implementation detail (I *think*?). How do we know we have to catch `json.JSONDecodeError` and not, like, `some_fancy_json.JSONDecodeError`?

mcous replied:
> First, it turns out .parse_file() takes a path, not a file-like object. 🙃

🤦 of course it does

> In other words we can't get rid of the JSONDecodeError. Pydantic doesn't appear to convert it to PydanticValidationError; it lets it propagate.
>
> Catching it like this felt fragile to me because Pydantic's choice of default JSON deserializer is an undocumented implementation detail (I think?)

I think Pydantic propogating the `JSONDecodeError` is probably a good thing (or at least, a not horrible thing). We _should_ differentiate between something that isn't JSON and something that is not the JSON document we expect.

Also, Pydantic's docs [mention that it uses json.dumps explicitely](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson). In the realm of "things we can rely on", I think Pydantic using Python's built-in `json` module as the default json implementation is pretty rock solid. Also, it's not like the version of Pydantic we're using is going to change under our feet.

I would probably write this something like this:

```python
def _analyze_json(main_file: InputFile) -> "JsonPreAnalysis":
    try:
        # at this point both _analyze_json and _analyze_python would be
        # calling file_like.read(), so maybe move that (and a seek(0))
        # external to this function
        contents = main_file.file_like.read()
        protocol = JsonProtocol.parse_raw(contents)
    except IOError:
        # ...handle read error
    except JSONDecodeError as exception:
        raise JsonParseError() from exception
    except PydanticValidationError as exception:
        raise JsonSchemaValidationError() from exception

    return JsonPreAnalysis(metadata=protocol.metadata.dict())
```

It's short, it catches reasonable errors with only one ugly try/raise, and it leans on something else (Pydantic) to do all the hard stuff

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+    # We extract metadata directly from the JSON dict,
+    # instead of from the fully parsed Pydantic model.
+    # This way, we preserve metadata fields that the Pydantic model doesn't know about.
+    # (Otherwise, Pydantic would silently discard them.)
```
Can we set the `Metadata` class's config to allow extra properties instead? This will line the Pydantic model up with the JSON schema more accurately

SyntaxColoring replied:
Yes, done.

* `class Config: Extra.allow` was set on other parts of those Pydantic models, just not `Metadata`. Not sure why.
    * A problem with the JSON protocol schema?
    * A problem with autogenerating Pydantic models from the JSON protocol schema?
    * An oversight in manual fixups?
* Looking at this more closely made me realize some other gotchas: https://github.com/Opentrons/opentrons/issues/8334#issuecomment-921206045

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+
+@dataclass(frozen=True)
+class InputFile:  # noqa: D101
+    # todo(mm, 2021-09-14): Specify whether this is allowed to contain path separators,
```
After thinking about it, I don't think this is necessary. Let's add a docstring that specs it as the filename with suffix, to match `Path.name`

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+
+
+@dataclass(frozen=True)
+class InputFile:  # noqa: D101
```
I'm not sure I see the value of this dataclass at the moment, given that it's an arbitrary (looking) renaming of the fields from `starlette.datastructure.UploadFile`. Can we use a `typing.Protocol` here instead and go with structural subtyping for now?

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+
+
+@dataclass(frozen=True)
+class JsonPreAnalysis:  # noqa: D101
```
This class will need `schema_version` to support #8339

mcous commented on api/src/opentrons/protocols/models/json_protocol.py:
```diff
@@ -680,3 +680,29 @@ class Model(BaseModel):
     )
+
+
+def make_minimal(
```
This is only used by tests. I think it should be moved back to where it was

SyntaxColoring replied:
Right, but it's now used by tests in both `api` and `robot-server`.

mcous replied:
I would duplicate the helper in that case

SyntaxColoring replied:
Done, but this is surprising to me and I'd like to hear your thoughts.

SyntaxColoring replied:
[Discussed in call] Duplicated test fixtures feel better than a dead code path in production source.

mcous commented on robot-server/robot_server/protocols/dependencies.py:
```diff
@@ -43,6 +44,11 @@ def get_analysis_store(app_state: AppState = Depends(get_app_state)) -> Analysis
     return analysis_store
 
 
+async def get_pre_analyzer(app_state: AppState = Depends(get_app_state)) -> PreAnalyzer:
```
Unless there's an actual need for a factory, I woudn't add one

SyntaxColoring replied:
Could you clarify this comment?

The endpoint handler `create_protocol()` has a `pre_analyzer` argument so that the PyTest function `test_create_protocol` can replace it with a mock. For that argument to be a real `PreAnalyzer` in production, it needs to be a FastAPI dependency, right?

mcous replied:
To flesh out, I meant: "since `PreAnalyzer` is not a singleton nor does it have any dependencies it itself needs injected, then we're free to pass the `PreAnalyzer` constructor directly into `Depends` without the need to make a `get_pre_analyzer` function"

```python
async def create_protocol(
    # ...
    pre_analyzer: PreAnalyzer = Depends(PreAnalyzer),
    # ...
) -> ResponseModel[Protocol]:

SyntaxColoring replied:
Oh! Cool.

mcous commented on robot-server/robot_server/protocols/protocol_models.py:
```diff
@@ -1,23 +1,71 @@
+        ...,
+        description=(
+            "When this protocol was *uploaded.*"
+            " (`protocolMetadata` may have information about"
```
```suggestion
            " (`metadata` may have information about"
```

mcous commented on robot-server/robot_server/protocols/protocol_models.py:
```diff
@@ -1,23 +1,71 @@
+        ),
+    )
+
+    # todo(mm, 2021-09-07): Should this be exposed to clients? We don't expose Python
```
Yes it should, and we'll need to expose API level at some point. Generally, look at the current app's file upload page for answers to the question "what information do we need?"

SyntaxColoring committed:
Merge branch 'edge' into pe_respond_with_metadata

Conflicts:
* robot-server/robot_server/protocols/dependencies.py
* robot-server/tests/protocols/test_protocol_analyzer.py
* robot-server/tests/protocols/test_protocols_router.py

mcous commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -70,20 +83,50 @@ async def create_protocol(
+        # todo(mm, 2021-09-14): Should this be HTTP 422?
+        raise ProtocolFileInvalid(detail=str(e)).as_error(status.HTTP_400_BAD_REQUEST)
+
+    # Pre-analysis may have read files.
+    # Reset them so that if protocol_store.create() reads them again,
+    # it will correctly start from the beginning.
+    for file in files:
+        await file.seek(0)
```
I think it's the responsibility of `pre_analyzer.analyze` to ensure that it doesn't mutate its arguments in this way. Having this in the route handler is not collaborator work

SyntaxColoring replied:
Replying in https://github.com/Opentrons/opentrons/pull/8274/files#r710118492.

mcous commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -70,20 +83,50 @@ async def create_protocol(
+    ) -> PreAnalysisInputFile:
+        underlying_file = upload_file.file
+
+        # Starlette clearly documents this as a SpooledTemporaryFile, but currently
+        # (Starlette v0.13.2), mypy infers Union[SpooledTemporaryFile[bytes], IO[Any]].
```
Might be the difference between `fastapi.UploadFile` and `starlette.datastructures.UploadFile`

SyntaxColoring replied:
I think this actually got magically resolved by one of the `edge` merges into this branch. A mypy update, I guess? Who knows. I hope to find peace with its dreadful mysteries.

mcous commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -70,20 +83,50 @@ async def create_protocol(
+    pre_analysis_input = [convert_upload_file_to_pre_analysis_input(u) for u in files]
+
+    try:
+        pre_analysis = pre_analyzer.analyze(pre_analysis_input)
```
Would be nice to move this down into the `try` block below

SyntaxColoring replied:
[Discussed in call] leaving this alone to keep the `try` blocks small and specific.

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+    # todo(mm, 2021-09-14): Specify whether this is allowed to contain path separators,
+    # and consider making it a pathlib.PurePath.
+    #
+    # Pre-analysis can read `file_like`. If, after pre-analysis, you want to read
+    # `file_like` again, it's your responsibility to call `file_like.seek(0)` to reset
+    # it.
```
This requirement feels pretty unintuitive for someone who might come in to use this interface, especially given that this file is an input argument

SyntaxColoring replied:
Yeah.

I thought about this, but came to the conclusion that this is a losing battle in Python; this is just how iterables and files work. I think it would be unconventional and inconsistent for the `PreAnalyzer` to take this responsibility.

As a general rule in Python, whenever you're reusing an iterable, it's your responsibility as the caller to reset it between uses, if you need to. Contrived example:

```python
dummy_data = [1, 2, 3, 4, 5]
my_iterable = reversed(dummy_data)
print(list(my_iterable))  # Prints [5, 4, 3, 2, 1].
print(list(my_iterable))  # Doesn't print anything; iterable was already exhausted.
```

File-like objects work the same way.

A less contrived example:

```python
file = open("Hello World.txt")
print(requests.post("http://httpbin.org/post", files={"my_file": file}).json())  # OK.
print(requests.post("http://httpbin.org/post", files={"my_file": file}).json())  # Wrong: file needs to be reset.
```

mcous replied:
Blah. Ok, if this is just too hard™️ at the moment, can we add a TODO comment to move to the "put the file on disk as part of pre-analysis" architecture that you drew out this week? Maybe that also resolves our naming issue: `ProtocolSaver` or something

mcous commented on api/src/opentrons/protocol_runner/pre_analysis.py:
```diff
@@ -0,0 +1,176 @@
+# todo(mm, 2021-09-13): Currently wrapped in a class so dependency-injection and mocking
+# feel less fragile, pending a team discussion about whether we still think this is
+# a good thing to do in Python.
+class PreAnalyzer:  # noqa: D101
```
Across the board, can we replace these `noqa: D101` with docstrings?

mcous commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -70,20 +83,50 @@ async def create_protocol(
+        # (Starlette v0.13.2), mypy infers Union[SpooledTemporaryFile[bytes], IO[Any]].
+        assert isinstance(underlying_file, SpooledTemporaryFile)
+
+        return PreAnalysisInputFile(upload_file.filename, underlying_file)
```
In general, can we use named parameters for dataclasses?

mcous commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -70,20 +83,50 @@ async def create_protocol(
-    if len(files) > 1:
-        raise NotImplementedError("Multi-file protocols not yet supported.")
+
+    def convert_upload_file_to_pre_analysis_input(
```
I'd be down to get rid of this function. I think it makes the whole thing harder to read and I don't see what value the `assert` is bringing to the table

SyntaxColoring replied:
The important part was the `assert`. It showed mypy that `underlying_file` was a `SpooledTemporaryFile[bytes]` and not an `IO[Any]`.

Like a `typing.cast()`, except with runtime typechecking also, because I wasn't 100% confident that I wasn't missing something. And if it turned out that I was missing something, I would rather have an `AssertionError` here than a `AttributeError`/`ValueError`/`TypeError` somewhere down the line.

But anyway, this is all moot now because this function got removed per another discussion: https://github.com/Opentrons/opentrons/pull/8274#discussion_r710115697

mcous commented on robot-server/tests/protocols/test_protocol_store.py:
```diff
@@ -1,12 +1,16 @@
 import pytest
 from datetime import datetime
-from decoy import matchers
+from decoy import Decoy, matchers
```
Unused `Decoy` should be removed

SyntaxColoring replied:
Nice catch, thanks.

mcous reviewed (changes requested):
Looking pretty good! I think we're doing a little too much in the create protocols route handler, and I've got assorted thoughts

SyntaxColoring committed:
Remove unused `Decoy` import.

SyntaxColoring committed:
Remove unnecessary `UploadFile.file` type assertion.

SyntaxColoring committed:
Fix docs: `protocolMetadata` -> `metadata`

Co-authored-by: Mike Cousins <mike@opentrons.com>

SyntaxColoring committed:
Add docstrings to `InputFile`.

SyntaxColoring committed:
Remove todo about not exposing `protocolType`.

Per review, yes, we should expose this to clients.

SyntaxColoring committed:
Make small revisions to `Metadata` model.

* Use Pydantic's `Extra.allow` enum instead of `"allow"` string literal.
* Clarify comment and turn it into a docstring.
* Leave todo for checking whether a dict[str, Any] would be better.

SyntaxColoring committed:
Delete unnecessary `get_pre_analyzer()` factory.

SyntaxColoring committed:
Fix whitespace.

SyntaxColoring committed:
Partially resolve `test_protocol_store` todos about input files.

SyntaxColoring committed:
Fix up and reinstate `test_get_all_protocols()`.

SyntaxColoring committed:
Add todo for `pathlib.Path` refactor.

SyntaxColoring committed:
Avoid raw JSON extraction hack by allowing extra fields in model config.

SyntaxColoring committed:
Duplicate `_make_minimal_json_protocol()` across project tests.

* Revert addition of `make_minimal()` to `models.json_protocol`:
  * Revert commit 579c1d06ff04f22c0517037fe29deb454ff279a0.
  * This reverts commit 2e0152588b7f03a6e7a79b5a99c73913db96d5ac.
* Add mostly copy-pasted function to `test_pre_analysis.py`.

mcous commented on robot-server/robot_server/protocols/protocol_models.py:
```diff
@@ -1,23 +1,72 @@
 
 
+class Metadata(BaseModel):
+    """Extra, nonessential information about the protocol.
+
+    This can include things like:
+
+    * A human-readable title and description.
+    * A last-modified date.
+    * A list of authors.
+
+    The exact set of fields is not defined by this API.
+    A protocol can define whatever it likes as its metadata,
+    and this API will pass it along without modification.
+    When reading a protocol's metadata,
+    you should treat it as an arbitrary JSON object from an untrusted source.
+
+    This metadata does *not* include
+    anything material to how the robot will run the protocol.
+    For example, it does not include the protocol's pipette and labware requirements.
+
+    (Exception:
+    Python protocols define their `apiLevel`,
+    which *does* affect how they're run,
+    as part of their metadata.
+    This is for historical reasons;
+    `apiLevel` is not truly metadata and doesn't really belong here.
+    You should not rely on a Python protocol's correct `apiLevel` being present here.
+    If a compelling need arises, this API may expose `apiLevel` some other way.)
+    """
```
I might consider paring this down. Detail is good, but brevity is also really important to make sure people can (and will) read it. Also, given discussions today about #8334, I don't think it's a good idea to commit to `The exact set of fields is not defined by this API.` right now

```suggestion
    """Extra, nonessential information about the protocol.

    This can include data like:

    * A human-readable title and description.
    * A last-modified date.
    * A list of authors.

    Metadata may contain fields other than those explicitly
    listed in this schema.

    The metadata *should not* include information needed
    to run the protocol correctly. For historical reasons, Python
    protocols define their `apiLevel` inside their metadata, but
    this should be considered an exception to the rule.
    """
    # todo(mm, 2021-09-17): Revise these docs after specifying
    # metadata more. github.com/Opentrons/opentrons/issues/8334
```

SyntaxColoring committed:
Replace some `noqa`s with docstrings.

mcous commented on api/src/opentrons/protocols/models/json_protocol.py:
```diff
@@ -77,6 +77,9 @@ class Metadata(BaseModel):
     Optional metadata about the protocol
     """
 
+    class Config:
+        extra = Extra.allow
```
👍

SyntaxColoring committed:
Add `schema_version` to `JsonPreAnalysis`.

SyntaxColoring committed:
Improve `Metadata` docstring.

Co-authored-by: Mike Cousins <mike@opentrons.com>

SyntaxColoring committed:
Style: Write out PreAnalysisInputFile kwargs.

SyntaxColoring committed:
Refactor `PreAnalyzer`.

* Make JSON and Python analysis functions take bytes instead of a file.
* Simplify JSON parse error handling.
* Style fixups.

SyntaxColoring committed:
Fix failed JSON parsing raising the wrong exception type.

mcous reviewed (approved):
🤘
